### DISCUSSION TITLE
How to have a `pagetemplate` last longer than a page?

### DISCUSSION THREAD

**Discussion by AzuxirenLeadGuy** [sile-typesetter/sile#1919] [Q&A] [answered]
I am using the page template described in the SILE manual. The described page template works only as the first page, and the the default document layout seems to take over from the next page onwards.

Here is the document I am using

```
\begin[papersize=a5]{document}
\script[src=packages/lorem]
\par

\begin[first-content-frame=leftCol]{pagetemplate}

\frame[id=gutter, width=3%pw]

\frame[id=leftCol, left=left(content), right=left(gutter),
top=top(content), bottom=bottom(content),
next=rightCol]

\frame[id=rightCol, left=right(gutter), right=right(content),
top=top(content), bottom=bottom(content),
width=width(leftCol)]

\end{pagetemplate}

\lorem[words=3000]

\end{document}
```

On compilation, the described layout is only present for the first page, while the default layout is present for the remaining of the 6 page document.

**Comment by Omikhleia** [comment, marked as answer]
Greetings

> The described page template works only as the first page

I think that it is what it does _by definition_ = changing the page template on the _current_ page.

The `\pagetemplate` command is a fairly low-level construct used in some advanced use cases.

I cannot say without more context, but you might actually be using the wrong tool -- It seems to me that the **masters** package is what you might be looking for. It allows to define sets of frames (`\define-master-template`) and switch between them either temporarily or permanently (resp. `\switch-master-one-page` and `\switch-master`).

Example:

```
\use[module=packages.masters]
% define a named master template 'mymaster'
% It is patterned on the \pagetemplate command
\define-master-template[id=mymaster,first-content-frame=content]{
  \frame[id=content, ... your definition ...]
  \frame[id=folio, ... your definition ...]
 ... other frames possibly ....
}
% Switch to that master
\switch-master[id=mymaster]
```

(_Unrelated_, but we should put a deprecation warning on things such as `\script[src=packages/lorem]`, it's weird to still see it used vs. the cleaner `\use[module=packages.lorem]` introduced more that 1 year ago...)

**Reply by AzuxirenLeadGuy** [reply]
Thank you so much. You perfectly answered what I needed.

**Reply by alerque** [reply]
We didn't add a deprecation notice to `\script[src=...]` because we were not removing it entirely, only de-conflating the possible uses. Anything that loads a known SILE module tipe (class, package, inputter, outputter, etc.) should use the `\use[module=...]` syntax, but just throwing a raw Lua script at the wall is still supported. Personally I would suggest all cases of the latter migrate to `\lua[require=...]` instead. I'll look into detecting specific cases and outputting a deprecation notice that points in the right direction.

**Reply by Omikhleia** [reply]
@alerque or a recommendation, if a deprecation is felt too strong? The documentation is lacking a bit here, however (#1854)

**Reply by alerque** [reply]
Not being sure whether this should be a recommendation or a deprecation is why it never got done. I'm still open to considering that point. I'm working on a PR and I think we can give pretty solid recommendations for most use cases and I'm leaning towards it being a deprecation in 0.15.0  with removal in 0.16.0 rather than just being a suggestion.

**Reply by alerque** [reply]
c.f. #1922, lets move discussion on this point there.

**Reply by davidchisnall** [reply]
I'm trying to follow this example, but every variant I come up with fails.  Is there a complete example of a two-page layout somewhere?

**Reply by AzuxirenLeadGuy** [reply]
@davidchisnall 
> I'm trying to follow this example, but every variant I come up with fails. Is there a complete example of a two-page layout somewhere?

Hi. My original question was about using page templates, and the answer was to use the `masters` package to create page templates, and switching them. This package is "not standard", so that explains why you don't see much documentation on it. You will see a brief explanation of it in the Manual PDF in section 5. There is also a single example as mentioned in the PDF at [`tests/masters.sil`](https://github.com/sile-typesetter/sile/blob/master/tests/masters.sil) and I do feel that the documentation is lacking. 

You can view my example that worked for me, and it I have commented it to explain it. Even then, I found the settings of the page template to be fragile, so it took me (at that time) a lot of trials to get it working correctly. I don't use SILE these days, so it might be better now. Hope it helps.

```
% Start writing this document on A5 page
\begin[papersize=a5]{document}

% Use the script 'packages/lorem'. This scirpt is used to generate text for content.
\script[src=packages/lorem]

% Use the module 'packages.masters'. This package contains the logic for making custom page templates
\use[module=packages.masters]

% define a named master template 'mymaster_a'
\define-master-template[id=mymaster_a,first-content-frame=t]{
	% Here, this template has the id "mymaster_a", and the text is writted
	% from the content frame with id "t"

	
	\frame[id=gutter,width=10%pw]
	% A frame is defined with width of 10% the width of page


	\frame[id=t,left=5%pw,right=95%pw,top=10%ph,bottom=35%ph,next=l]
	% A frame is defined with id='t', with the given left and right x-coordinates, top and bottom y-coordinates. 
	% If the text content in this frame is filled, the text should flow to next frame id='l'

	
	\frame[id=l,left=5%pw,right=left(gutter),top=bottom(t)+2em,bottom=90%ph,next=r]
	% A frame is defined with id='l', with the given left and right x-coordinates, top and bottom y-coordinates. 
	% If the text content in this frame is filled, the text should flow to next frame id='r'

	\frame[id=r,left=right(gutter),right=95%pw,top=top(l),bottom=bottom(l),width=width(l)]
	% A frame is defined with id='r', with the given left and right x-coordinates, top and bottom y-coordinates. 
	% If the text content in this frame is filled, the text should flow to the next page (since no next is specified)
}


% define a named master template 'mymaster_b'
\define-master-template[id=mymaster_b,first-content-frame=leftCol]{
	% Here, this template has the id "mymaster_b", and the text is writted
	% from the content frame with id "leftCol"

	\frame[id=gutter, width=10%pw]

	\frame[id=leftCol,left=5%pw,right=left(gutter),top=5%ph,bottom=95%ph,next=rightCol]
	
	\frame[id=rightCol,left=right(gutter),right=95%pw,width=width(leftCol),top=top(leftCol),bottom=bottom(leftCol)]
}



\nofolios
% This command disables putting page number/footer details, since
% it would mess with our templates.  

\switch-master[id=mymaster_a]
% Switch to that master


\lorem[words=1600]
% Generate 1600 random words. About 3 pages in my PC.

\framebreak
\framebreak
\supereject

\switch-master[id=mymaster_b]
% Switch to that master

% Need to break pages so that the new master is initialized
% from the next page, otherwise, sile merges both templates 
% in the same page. I tried a lot before setting on a combo 
% that works

\lorem[words=1500]
% Generate 1500 random words. About 3 pages in my PC.

\end{document}
% End of the document

```

**Reply by Omikhleia** [reply]
Thanks @AzuxirenLeadGuy for the minimal example!
What I can add is that some packages and classes expect certain frames to exist: typically the "folio" as you noted; the book class also needs a frame for running headers; etc. (my own _re·sil·ient_ module has even more expectations on how frames must be minimally named and organized, for instance.)
So things may become "a bit" more complex when one wants to play well with these packages and classes... (in other terms, pagetemplate is very low level (and rather hard to use right, one really has to know what one does); page masters are higher-level (and a bit easier for regular use); and other modules may add their own logic on top of these (e.g. in _re·sil·ient_, I have "layouts", but of course an "easier" high-level abstraction also implies it's also less flexible for some use cases).
